src/clean-core/string_stream.hh

```diff
@@ -0,0 +1,109 @@
+#pragma once
+
+#include <clean-core/string.hh>
+#include <clean-core/string_view.hh>
+#include <cstring> // std::memcpy
+
+namespace cc
+{
+class string_stream
+{
+public: // methods
+    string_stream& operator<<(string_view sv)
+    {
+        reserve(sv.size());
+        std::memcpy(m_curr, sv.data(), sv.size());
+        m_curr += sv.size();
+        return *this;
+    }
+
+    [[nodiscard]] string to_string() const
+    {
+        string s = string::uninitialized(size());
+        memcpy(s.data(), m_data, size());
+        return s;
+    }
+
+    /// reseve space for at least size more elements
+    void reserve(size_t size)
+    {
+        size_t req_size = this->size() + size;
+        if (req_size <= m_capacity)
+            return;
+
+        size_t new_cap = (m_capacity << 1) < req_size ? req_size : (m_capacity << 1);
+        char* new_data = new char[new_cap];
+        std::memcpy(new_data, m_data, this->size());
```

Comment:
this is UB if the stringstream is default-initialized

Reply:
good point.

> If either dest or src is an invalid or null pointer, the behavior is undefined, even if count is zero.